### ISSUE TITLE
Need of x11

### ISSUE THREAD

**Issue by kinow** [BSC-ES/autosubmit#544]
In GitLab by @orioltinto on Jun 12, 2020, 12:42

In a workflow I'm working on we have some jobs that require to have a display defined and to have -x11 enabled even there's anything for the user to do interactively.

A simple test to reproduce the problem is to execute a simple template with:
`xclock`

I think that it should be possible to do it through paramiko.

**Comment by kinow** [comment]
In GitLab by @mcastril on Jun 12, 2020, 13:04

I could launch x11 inside a batch job with this headers:

```
#SBATCH --qos=bsc_es
#SBATCH -A bsc32
#
#
#
#SBATCH --cpus-per-task=1
#SBATCH -n 48
#SBATCH -t 00:05:00
#SBATCH -J a2wz_REMOTE_SETUP
#SBATCH --output=/gpfs/scratch/bsc32/bsc32252/a2wz/LOG_a2wz/a2wz_REMOTE_SETUP.20200612123431.out
#SBATCH --error=/gpfs/scratch/bsc32/bsc32252/a2wz/LOG_a2wz/a2wz_REMOTE_SETUP.20200612123431.err
#SBATCH --x11=batch
```

The important one for this case is `--x11=batch`

This was run after checking that I could run `xclock` from the login terminal. The problem, as @orioltinto accurately pointed out, it seems it is that Paramiko is not correctly forwarding X11.

I saw some messages regarding this matter, so maybe it is possible:

https://stackoverflow.com/questions/12354047/x11-forwarding-with-paramiko

@dbeltrankyl @wuruchi

**Comment by kinow** [comment]
In GitLab by @mcastril on Jun 12, 2020, 13:32

If you want to test you can copy my experiment a2wz. The template is in the same experiment (I used local, creating the directory in the experiment itself).

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Jul 23, 2020, 10:34

Hi

I've worked on this but so far I didn't get satisfying results.

Paramiko is not able to use x11 forwarding at ssh client level ( at least I didn't found it), instead it has to create a new session on the transport layer and in this layer, it can request an x11 connection. 
This x11 connection is not implemented on paramiko itself it has to be added manually through a handler using the Xlib library, so essentially paramiko only adds the possibility of establishing a connection.

After adding manually a handler for x11 (using xlib), you can use simple commands like exec_command("xclock") however it only will work though direct commands like this one since the connection will be closed once the command finishes since it is done on a session-level( alias one command).

Due to the nature of job submission, the issue is that paramiko can't know that there will be an x11 request somewhere in the future when the job begins.  

So the main issue resides on that paramiko can't forward ssh connections of slurm batch since the session will be closed

So in summary, the main issue resides on the way that autosubmit launches a job though paramiko. It creates a connection with exec-command and this connection only includes the submission of the jobs(multiple at once) not the execution of the command.

So the solution could be that paramiko perma-open an x11 channel and receive all x11 connections from the cluster though a thread on autosubmit, however as I said, I don't see that this possibility is implemented on paramiko library.

 When @mcastril returns from holidays we will discuss maybe there are something that I didn't find.

Edit: Also this connection is always established from paramiko to the cluster, there is no way to "wait for connection" from cluster to paramiko.

**Comment by kinow** [comment]
In GitLab by @mcastril on Aug 10, 2020, 12:58

In the meeting last week we discussed two possibilities:

* Try using the interactive partition. An example is the COMPILATION step in Auto-EC-Earth workflow.
* Try declaring a PS machine, which runs things directly in the login. This is not the best solution but at least will help to know if it is possible to achieve an X11 connection through Autosubmit.

**Comment by kinow** [comment]
In GitLab by @mcastril on Oct 1, 2020, 13:47

Oriol, did you look if it was working using -p interactive? I don't remember if we did the test.

**Comment by kinow** [comment]
In GitLab by @mcastril on Oct 15, 2020, 17:25

mentioned in issue ces/nemo_performance_profile#41

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on May 13, 2021, 13:48

Hello @mcastril , 

I have a working demo, with a very reduced version of autosubmit ( only submit a template get jobid and queue status). I only tried to display xclock and a xterm. 

I tried two schemes

* Using SRUN, since it is an interactive command thus paramiko won't close the channel until the srun script finishes, it works in paramiko adapting a bit the work done previously, however, there is a big downside for use srun within autosubmit since I would need to change so many things for run autosubmit with srun. 

* With sbatch, since it is not interactive, paramiko just close the channel when it gets the EOF of the job submitted. However, I did a bit of hack and launched the command with a (sbatch -D job_script && sleep infinity) this keeps the channel open forever and thus the X11 connections can be redirected to our local display successfully.

The sbatch solution has the issue that the channel is blocking, but with threads and with a SACCT command ( if job is completed => close the channel) I can close the channel manually and avoid the block.

I will implement the sbatch solution demo into autosubmit code.

For now, I adapted the work done with x11 to 3.14 version, and completed the demo with threads. So it is matter of time

**Comment by kinow** [comment]
In GitLab by @mcastril on May 13, 2021, 13:53

Congratulations @dbeltrankyl! I think it's a very smart solution. It's clear that there was not an easy way to deal with this. As it will only be used in very specific situations (such as launching Paraver) I would minimize the impact on Autosubmit code, so I think that option 2 is the wisest as you suggest.

**Comment by kinow** [comment]
In GitLab by @mcastril on May 13, 2021, 16:51

What do you think? Would it be ok for 3.14.0? Or we wait for AS4. Another option is to do a transitional version between AS3 and AS4 with this kind of experimental improvement needing additional dependencies.

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on May 14, 2021, 09:27

I think it will be ok, 

I didn't notice any issue so far. In any case when I begin the test, I'll have more feedback regarding this

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Jul 22, 2021, 13:00

Hello, 

This feature is now included in the 3.14.0-beta release that will happen when the maintenance finishes.

To active this feature, you only will have to put 
```
[job_section]
FILE = ...
...
X11 = true
```

NOTE: It is not available for wrappers and only will work under slurm schedule.

**Comment by kinow** [comment]
In GitLab by @ macosta on Jul 23, 2021, 12:54

Nice job! Probably is a good moment now that @dbeltrankyl is working with the BSC Tools. Could we organize a meeting next week to coordinate some work of the auto-perf? 

@mcastril

**Comment by kinow** [comment]
In GitLab by @mcastril on Jul 23, 2021, 12:58

Ok, Wednesday seems the best day, taking into account the availability of everybody. I will send an invitation @dbeltrankyl @sparonuz @xyepes

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Jul 23, 2021, 13:15

It's fine for me, thanks!

**Comment by kinow** [comment]
In GitLab by @sparonuz on Jul 23, 2021, 14:33

for me too

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Jul 29, 2021, 14:43

Hello, 

I tried to use paraver within this X11 implementation. 

* In the code isolated that I had in a demo. 
* Within Autosubmit. 

The job runs without issues but the graphical screen doesn't show up as it does with xclock. 

I think that perhaps there is an issue triggered when using EXPORT=module load paraver; since it is added into the command itself. 

I will test with other configurations as soon as possible

**Comment by kinow** [comment]
In GitLab by @rocsalvadorbsc on Aug 3, 2023, 12:31

Hi,

I know this issue was stopped a long ago, but now since I am porting the automatic performance profiling tool to autosubmit and it uses Paraver, I need to use X11. I don't know if there's any possibility for it to work in Marenostrum 4 or if I should wait for Marenostrum 5.

Thank you

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Aug 3, 2023, 13:58

Hello, 

In as3 it worked for a xclock, but never managed to run paraver not sure exactly why but it was an issue from paramiko itself ( the lib that we use to send jobs ) 

I will try to reenable the X11 and see if the paramiko has now better support for it.

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Aug 3, 2023, 14:32

Seems that works for local but... for mn4 I have bad news


I just remind this mail from 21/06/22

```
Dear users,

Slurm has been successfully upgraded on MN4.

Command sbatch does not support x11 directive, please, use salloc or srun instead.

In case any issue arises, please, do not hesitate to contact us.

Thanks for your understanding.

BSC Support Team.
```

This is a big issue because the actual code relies upon using --x11=batch .. all jobs are submitted using sbatch not sure if a workaround is possible, FYI @mcastril

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Aug 3, 2023, 14:55

I guess I should prompt a salloc instead of sbatch when dealing with X11

current issues without thinking much:

* salloc doesn't understand #sbatch directives -> Need to make an auto translator into command line options for all the directives on the cmd... 
* salloc is interactive -> Need to encapsulate it into a thread 
  this is tricky.. as Autosubmit expects the sbatch behavior, luckily it returns a jobid instantaneous so it seems doable

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Aug 3, 2023, 16:09

Miguel told me that he will ask to David Vicente

**Comment by kinow** [comment]
In GitLab by @rocsalvadorbsc on Aug 8, 2023, 10:49

mentioned in issue ces/automatic_performance_profiling#41

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Aug 30, 2023, 08:35

Hello @rocsalvadorbsc ,

Are you connected to the AS machine with ssh -XY?

Can you open a terminal xclock or gedit or any GUI app? to ensure that it is not an issue with the module

**Comment by kinow** [comment]
In GitLab by @rocsalvadorbsc on Aug 30, 2023, 08:58

Hi @dbeltrankyl,

I was running it directly from the workstation with no ssh. Now, I tried to run it with `ssh -XY` from my laptop (gedit opens with no issue) but I get this error when the script reaches the paraver command:

`X11 connection rejected because of wrong authentication.`

**Comment by kinow** [comment]
In GitLab by @albertvilabsc on Aug 30, 2023, 11:48

Hi @dbeltrankyl ,

Can you send me the intructions to reproduce this in the WS and autosubmit VM?

Many thanks,

Albert

**Comment by kinow** [comment]
In GitLab by @albertvilabsc on Aug 30, 2023, 13:26

@dbeltrankyl I will make some tests soon. In the bsceshub it is working?

Thanks,

Albert

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Aug 30, 2023, 13:41

I did not test it in the bsceshub only VM and office ones (and in my laptop in which it is working ) but I can upload it and test it

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Aug 30, 2023, 15:18

Hello,

I've uploaded it to bschubs 

In case of permission issues in my /home/Earth/dbeltran, perform 

`cp -r /esarchive/autosubmit/a6en/proj /esarchive/autosubmit/$expid/.`

Before the create.


I've tested it, and the same issue is rising.

```bash
timeout 156.0 nohup bash /gpfs/scratch/bsc32/bsc32070/a6en/LOG_a6en/a6en_PARAVER.cmd > /gpfs/scratch/bsc32/bsc32070/a6en/LOG_a6en/a6en_PARAVER.cmd.out.0 2> /gpfs/scratch/bsc32/bsc32070/a6en/LOG_a6en/a6en_PARAVER.cmd.err.0 & echo $! ; sleep 156.0 2>/dev/null
X11 connection rejected because of wrong authentication.
```

**Comment by kinow** [comment]
In GitLab by @mcastril on Aug 30, 2023, 21:00

I guess that you already checked the $DISPLAY variable and etc.

We had a related discussion in the Docker MR:

https://earth.bsc.es/gitlab/es/autosubmit/-/merge_requests/364#note_227693

In the discussion, Bruno was pointing to this interesting [post](https://goteleport.com/blog/x11-forwarding/)

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Aug 31, 2023, 08:12

> In the discussion, Bruno was pointing to this interesting [post](https://goteleport.com/blog/x11-forwarding/)

Thanks, Miguel I'll look into it.

> I guess that you already checked the $DISPLAY variable and etc.

Autosubmit should be able to do it automatically, like on my laptop, but just in case, I manually exported these parameters, and it did not work.

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Jun 11, 2024, 11:22

4.1.7 is now available in the hubs, but I haven't had time to test it yet. I created a new issue, https://earth.bsc.es/gitlab/es/autosubmit/-/issues/1321, to continue the tests/ discussion.
